n8n-io/n8n
# fix(core): Filter out certain executions from crash recovery

## Summary

This PR fixes an unwanted interaction between concurrency control and crash recovery, by logging new events (execution throttled, execution started during bootup) and filtering out those executions from the log writer's check.

## Context

During startup we [start running](https://github.com/n8n-io/n8n/blob/16b1a094b19e5f803a460b99c6062a1175bec153/packages/cli/src/commands/start.ts#L304C15-L304C36) any executions that had been enqueued at the time of shutdown, running them up to the concurrency limit and enqueuing the excess. This does not block, so that concurrency control can let through or throttle these executions as usual.

During startup we also have the execution recovery service, which receives event logs about unfinished executions and uses them to amend data in executions truncated by an instance crash. It is the log writer that [collects](https://github.com/n8n-io/n8n/blob/16b1a094b19e5f803a460b99c6062a1175bec153/packages/cli/src/eventbus/MessageEventBusWriter/MessageEventBusLogWriter.ts#L207) these event logs about unfinished executions, by checking for events written to `n8nEventLog-{n}.log`. 

Hence during startup we move executions to `running` status up to the limit and keep the excess with `new` (i.e. enqueued) status, but because all these executions have not finished yet, the log writer reports them to the recovery service, which marks them as crashed. 

## Testing

Currently the log writer is not easily testable. To test manually:

- Ensure you have no `new` or `running` executions. Set `N8N_CONCURRENCY_PRODUCTION_LIMIT=2` and start the instance.
- Create a workflow with a 2s cron and a 20s wait, activate it. Allow 2 `running` executions and a handful of `new` executions to accumulate.
- Deactivate the workflow, stop the instance. After graceful shutdown there should be 2 `success` and a handful of `new` executions. Start the instance again.
- Result - `new` executions are moved to `running` up to the limit, without being marked as `crashed`. 

## Related Linear tickets, Github issues, and Community forum posts

https://n8nio.slack.com/archives/C069HS026UF/p1719846780406959?thread_ts=1719825397.534179&cid=C069HS026UF

## Review / Merge checklist
